engines/order_management/lib/order_management/reports/enterprise_fee_summary/renderers/html_renderer.rb

```diff
@@ -0,0 +1,44 @@
+require "open_food_network/reports/renderers/base"
+
+module OrderManagement
+  module Reports
+    module EnterpriseFeeSummary
+      module Renderers
+        class HtmlRenderer < OpenFoodNetwork::Reports::Renderers::Base
+          def header
+            data_row_attributes.map do |attribute|
+              header_label(attribute)
+            end
+          end
+
+          def data_rows
+            report_data.enterprise_fee_type_totals.list.map do |data|
+              data_row_attributes.map do |attribute|
+                data.public_send(attribute)
+              end
+            end
+          end
+
+          private
+
+          def data_row_attributes
+            [
+              :fee_type,
+              :enterprise_name,
+              :fee_name,
+              :customer_name,
+              :fee_placement,
+              :fee_calculated_on_transfer_through_name,
+              :tax_category_name,
+              :total_amount
+            ]
+          end
+
+          def header_label(attribute)
+            I18n.t("order_management.reports.enterprise_fee_summary.formats.html.header.#{attribute}")
+          end
+        end
```

Comment:
This is looking very good but I still fill it's more coupled to the actual report details than about Html, which is the opposite of what we are aiming for.

https://github.com/davidgf/design-patterns-in-ruby/blob/master/strategy.md is exactly what I have in mind. We could name these `enterprise_fee_type_totals` and `list` methods more abstract and with a clear API that makes it a breeze to add a different formatter in the future.

Reply:
@sauloperez I see what you're saying. Are you okay with pairing the HTML renderer class with a HAML view, but one that is not tied with the controller? For example, a view in `app/views/order_management/reports/enterprise_fee_summary/html_renderer/report.html.haml` of the `order_management` engine. The directory structure would follow `String#underscore` of the renderer.

About using `enterprise_fee_type_totals` and not a more generic method name, those are in anticipation of downloadable reports that contain more than one set of data. For example, a hypothetical report may start with the $$ monthly summary for the year, the next pages to list nearing out of stock items, etc. So what I had in mind was:

1. `ReportService` - Integrates the different classes that are needed to build the report data and render the report
2. `ReportService#enterprise_fee_type_totals` - Represents one table in the report. There can be more!
3. `ReportData::EnterpriseFeeTypeTotals` - This is used by `enterprise_fee_type_totals`. We can make it a convention that table data classes use `#list`. Perhaps we should move the headers in the renderers here?

What do you think?

Reply:
I don't understand what you mean by, sorry

> The directory structure would follow String#underscore of the renderer.

I guess you mean that we'll pass the renderer object to the haml and it'll reference its methods?

Regarding anticipation, I would not do it. I would not anticipate until we have concrete requirements on the table. In my experience, it leads to YAGNI and the wrong abstractions, which are very costly.

My point was to have the common algorithm that both the CSV and HTML renderers follow implemented in the `Base` class. Now it's almost empty. All renderers should share the same duck-type/interface.

I wanted to share a snippet but I think I went too far :sweat_smile:. Do you fancy a videocall? I think it'll be quicker for both of us.

Reply:
> The directory structure would follow String#underscore of the renderer.

I just meant that, instead of having the HTML template in `spree/admin/reports/enterprise_fee_summary_report`, `OrderManagement::Reports::EnterpriseFeeSummary::Renderers::HtmlRenderer` owns it so we could have it in (`String#underscore`) `order_management/reports/enterprise_fee_summary/renderers/html_renderer`. But I've changed my mind!

As we discussed in Slack, I realized this HTML template would always be used by the controller, so I'm okay with using the current template path. For any future XLS/PDF/others report generation that will require a view template but ideally can be generated without a controller, we can decide what path to use when that is needed.

Reply:
> Regarding anticipation, I would not do it. I would not anticipate until we have concrete requirements on the table. In my experience, it leads to YAGNI and the wrong abstractions, which are very costly.
> 
> My point was to have the common algorithm that both the CSV and HTML renderers follow implemented in the Base class. Now it's almost empty. All renderers should share the same duck-type/interface.

I get your point. I will rename `#enterprise_fee_type_totals` to `#data`, and subclass `::EnterpriseFeeTypeTotals` to `::Table` (and probably remove the former altogether!).

Reply:
For this point, I removed `::EnterpriseFeeTypeTotals` altogether, and just used the method `list` returning an array in the report service. :+1: